app/services/invoices/payments/adyen_service.rb

```diff
@@ -0,0 +1,178 @@
+        update_invoice_payment_status(payment_status: invoice_payment_status)
+
+        result.payment = payment
+        result
+      end
+
+      def update_payment_status(provider_payment_id:, status:)
+        payment = Payment.find_by(provider_payment_id:)
+        return result.not_found_failure!(resource: 'adyen_payment') unless payment
+
+        result.payment = payment
+        result.invoice = payment.invoice
+        return result if payment.invoice.succeeded?
+
+        payment.update!(status:)
+
+        invoice_payment_status = invoice_payment_status(status)
+        update_invoice_payment_status(payment_status: invoice_payment_status)
+
+        result
+      rescue BaseService::FailedResult => e
+        result.fail_with_error!(e)
+      end
+
+      private
+
+      attr_accessor :invoice
+
+      delegate :organization, :customer, to: :invoice
+
+      def should_process_payment?
+        return false if invoice.succeeded?
+        return false if adyen_payment_provider.blank?
+
+        customer&.adyen_customer&.provider_customer_id
+      end
+
+      def client
+        @client ||= Adyen::Client.new(
+          api_key: adyen_payment_provider.api_key,
+          env: adyen_payment_provider.environment,
+          live_url_prefix: adyen_payment_provider.live_prefix
+        )
+      end
+
+      def adyen_payment_provider
+        @adyen_payment_provider ||= organization.adyen_payment_provider
+      end
+
+      def update_payment_method_id
+        result = client.checkout.payments_api.payment_methods(payment_method_params).response
+
+        if (payment_method_id = result['storedPaymentMethods']&.first&.dig('id'))
+          customer.adyen_customer.update! payment_method_id: payment_method_id
+        end
+      rescue Adyen::AdyenError => e
+        deliver_error_webhook(e)
+        raise
+      end
+
+      def create_adyen_payment
+        update_payment_method_id
+
+        client.checkout.payments_api.payments(payment_params).response
+      rescue Adyen::AdyenError => e
+        deliver_error_webhook(e)
+        update_invoice_payment_status(payment_status: :failed, deliver_webhook: false)
+
+        raise
+      end
+
+      def payment_method_params
+        {
+          merchantAccount: adyen_payment_provider.merchant_account,
+          shopperReference: customer.external_id,
+        }
+      end
+
+      def payment_params
+        prms = {
+          amount: {
+            currency: invoice.currency.upcase,
+            value: invoice.total_amount_cents
+          },
+          reference: invoice.number,
+          paymentMethod: {
+            type: "scheme",
+            storedPaymentMethodId: customer.adyen_customer.payment_method_id
+          },
+          shopperReference: customer.external_id,
+          merchantAccount: adyen_payment_provider.merchant_account,
+          shopperInteraction: "ContAuth",
+          recurringProcessingModel: "UnscheduledCardOnFile"
```

Comment:
Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.